cpp/src/arrow/engine/substrait/extension_set.h

```diff
@@ -226,14 +246,10 @@ class ARROW_ENGINE_EXPORT ExtensionSet {
  private:
   ExtensionIdRegistry* registry_;
   /// The subset of extension registry URIs referenced by this extension set
-  std::vector<util::string_view> uris_;
+  std::unordered_map<uint32_t, util::string_view> uris_;
   std::vector<TypeRecord> types_;
-
   std::vector<FunctionRecord> functions_;
-
-  // pimpl pattern to hide lookup details
-  struct Impl;
-  std::unique_ptr<Impl, void (*)(Impl*)> impl_;
+  std::unordered_map<Id, uint32_t, IdHashEq, IdHashEq> types_map_, functions_map_;
```

Comment:
Does this need to be exposed in the public header? The pImpl at least kept these details out of the header.

Reply:
Do we have a guideline for hiding.  I find it hard to know when details are "too messy" and deserve a pimpl (other than mutex which we have lint checks for).

Given that users of this header are going to have to dealing with `Id` (`ExtensionIdRegistry` exposes the idea) the `IdHashEq` could be considered useful.

Reply:
I don't think we have a real guideline. IdHashEq could indeed be useful, but unordered_map is a "heavy" include to have (that said we lost that battle a long time ago). I don't think we have to hide this, though, just wondering if it's useful.